### DISCUSSION TITLE
Remove features which doesn't provide enough value and clutter UI too much

### DISCUSSION THREAD

**Discussion by fm-sys** [cgeo/cgeo#15434] [General]
In #15425 @eddiemuc mentioned that routing GUI is too heavy-weighted, and that we should take steps to clean up the GUI in general. @ztNFny is privately using a "cleaned up" version of c:geo. And for my side, I honestly gave up c:geo's UI completely. Depending on my current mood, I either argue against new features or just let it happen.

Keep in mind: 
We are power users. Almost nobody knows c:geo better than we do, and yet even we feel overwhelmed by using it.

**I want to use this discussion to collect features which clutter c:geo's UI too much** and can be removed or placed in a more streamlined way.

---

Somehow related:

Already quite some time ago I wrote down some [UI conventions](https://github.com/cgeo/cgeo/wiki/Themes%2C-styles-and-other-conventions#ui-conventions), however they were not yet discussed by the team. I don't know if they currently make sense and follow the c:geo's philosophy, but maybe we can write down some guidelines we all agree with in one of the next team meetings.

**Comment by eddiemuc** [comment]
I think it is a great idea to discuss this. We should put it on today's agenda IMHO.

My personal point of view:
As you know I am not a GUI/UX guy. You don't want me to design visual things for you, trust me :-) Take the recent "group store lists in dialog" as an example: I was truly grateful for @ztNFny s support there, otherwise it would have looked ugly. I love this way of working!

I am also not personally intrigued by complex GUIs which "normal" users may find ugly or non-understandable. However, I deeply respect and understand the wishes of others to have that. I also grant that c:geo is in a terrible state UX-wise.

Thus I am open and willing to follow any concept or rules we give ourselves as a team. I am also willing to work migrate existing stuff to such a concept. Where I can only be of little to no help however is **creating** such a concept.

What is important to me: same rules for all. If a new feature shall be introduced same GUI rules apply to and are enforced for everything. Regardless of personal feature preferences.

**Comment by eddiemuc** [comment]
One rather technical idea to approach at least the problem of ridicuously long options menus: implement generic settings to let users customize which options they want to see where and which not. Look & Feel could be very similar to the already introduced "quick launch button" settings:
![image](https://github.com/cgeo/cgeo/assets/6909759/a4c0ec97-567c-4ea1-90f7-67be42881fcf)

But for ALL option menues, e.g.
* the ones on activitys (screenshot examples are from map and stored lists):
![image](https://github.com/cgeo/cgeo/assets/6909759/7b48170f-99ec-43ef-bd4e-cdab0d268e4b)
![image](https://github.com/cgeo/cgeo/assets/6909759/0ccbd107-5b11-4c13-ba00-ae75bf16f10b)
* long-tap context menues e.g. on map or waypoints or user:
![image](https://github.com/cgeo/cgeo/assets/6909759/cffc5bc4-1583-4fa1-99e0-1a97d5305a61)
![image](https://github.com/cgeo/cgeo/assets/6909759/0d244f50-b297-46e2-bfda-d26a42a0dff5)
![image](https://github.com/cgeo/cgeo/assets/6909759/184076d5-4655-48f1-a48d-943a02529b24)
![image](https://github.com/cgeo/cgeo/assets/6909759/91aac4ae-a68e-4432-8f8b-adff63cf8581)

This only solves part of the problem, but could be a bridge between "power users" needing very specific functions and "normal users" only needing base functions. It could also provide a means for "normal users" to become "power users" over time

**Reply by fm-sys** [reply]
> This only solves part of the problem, but could be a bridge between "power users" needing very specific functions and "normal users" only needing base functions. It could also provide a means for "normal users" to become "power users" over time

I don't like the idea of treating users as "normal" or "power users" because I don't think those distinctions really exist. At least not how we developers think of them ;)

Even though I might use more advanced features, I still prefer a simple UI.

It seems like saying "let's do it for power users" is just a way to avoid designing a carefully planned UI.

Putting features in "advanced options" might work temporarily, but it's not a real solution.

**Reply by ztNFny** [reply]
My first thought upon reading that was: https://xkcd.com/927/ "Let's add more settings to our too-many settings to make it less setting"

Making the UI user-configurable is a great feature - for die-hard power users. It's basically what I'm doing with my custom builds, just a tad more accessible. But it's something that only very knowledgable users will use.
This suggestion could be implemented in 2 ways, both of which will imho not work - let's try to look at it from the eyes of a "normal" (not power) user:
- Everything on by default (= today). The user will see lots of options, not know what to do with half of them - same goes for the settings menu where those options would be, he probably wouldn't even discover them. Even if he does, to make the informed decision whether to en/disable a feature he needs to know that feature first - which he doesn't, so he can't.
- Everyhing off by default. You have very clean menus, but then very quickly discover that all the stuff you want to do isn't possible because there's no button for it. For some things you want to do maybe you know that you want to do it, so you might stumble upon the setting to enable it. But there are many settings that you have no idea what they do and thus will never enable them, even though they might be very valuable  (let's take routing as an example because even you as a dev didn't know what it does - you wouldn't enable the option and thus never get to know the feature - and thus again will be unable to make the informed decision whether you want the feature).

Leaving it up to the user is the "the dev's who know the application best couldn't decide on what's important, so now they push the decision to the user who just wants to use the app".

**Comment by ztNFny** [comment]
I think "remove features" is wrong approach to the problems cgeo has. The value of c:geo is that it has many features. The problem it has is not the feature count but the complexity of the UI. These 2 points are related, but far from identical.

Let me try to explain with a simplified example: A flashlight app. This app has only 1 feature. The simple UI will have 1 big button: On/Off. Yet even such a simple app could have a complex UI, e.g. you could have a slider for brightness, a timer for auto-off, different blinking modes, color wheel, share to facebook when the app is used, ...
What you need to do is evaluate what's the core value of your app and concentrate the UI on that. For the average flashlight app the core value is "provide light" and what's needed for that is that On/Off button, so that's all that there should be.

Now unfortunately for us c:geo is far from a flashlight app. What's even worse - and this makes the situation pretty much impossible to resolve without causing pain - is that it's been complex for a very long time. This lead to people using c:geo in very different way - because it allows those degrees of freedom. So even trying to identify the core values of the app is a quite hard endavour - ask 4 people and you'll get 5 answers.

With that said let me come back to the original "remove features": I'm absolutely certain that there's not a single feature in c:geo that can be removed without upsetting a considerable part of the userbase - those users for whom that feature that's not relevant to you is very relevant for in their workflow.

So what can we do? Imho we should not do nothing, we should try to improve. But doing that we must try to minimize pains - and that to me means we shouldn't touch the feature set (because again, that's what makes c:geo so great - sorry, another interlude here: I know many people who years ago didn't use c:geo because it was laking many essential features. They used e.g. Locus instead. And just today one of those former "Locus for Geocaching"-users told me that c:geo now has every feature he needs for geocaching + everything he needs for hiking, so he doesn't even need "Locus for Hiking" anymore. So all those features are certainly delivering value! - now back to the point:)
We should aim to make those existing features more accessible. For me this means thinking about how can we make them easier to understand, easier to control and "more intelligent" - if the feature does what you need by default then you don't need 3 buttons to fine-tune its behaviour.

What would be needed to follow this approach? I don't think there's any way to avoid the classical product development workflow here (and hey, that workflow is established industry standard for a reason!): Gather data on how people use the app, what problems they try to solve (= collect requirements, write user stories), which features they use for that, suggest them alternatives (= evaluate approaches and refine them). That's a huge amount of effort for sure - and considering the wide range of c:geo users to do it proper we'd have to look at every story with at least several people. And what's for sure: This can not be done in text form in a forum!
Maybe as a first step we could start in the team calls and pick out one feature, discuss that - but for every feature it'd probably take 2-3 calls to get somewhere (and that's without actual coding), so don't expect a solution tomorrow!

----------------------

All what I said above is on a meta-level, because putting that in practice is a quite big challenge - for every single feature! But let me try it with one specific example - routing - because we had that discussion in the other thread.

What's the usecases (that I can think of right now - this is where more perspectives are needed):
- I'm somewhere and want to find 1 cache real quick, show me where this cache is.
- I want to do a big tour this weekend and find these 30 caches, show me the ideal path using existing paths (not as the crow flies) to all of them
- I'm on such a pre-planned tour and decided to take a quick detour, add this cache here
- I want to go hiking and find some caches along the way
- I want to by car to 1 specific cache, use car navigation (e.g. google maps) to get there
- I want to know the distance from A to B

So it seems we need at least "add navigation target", "add Xth target", "set navigation mode (as the crow flies, hiking, biking, car, ...)"
What do we have today:
- in the map context menu there are currently 4 routing related menu items: Append to route, Prepend to route, Set as target, Navigate.
- Also in a cache popup there's Compass, Set as target, Navigate,
- 3 "routing modes": target (red line), route (blue line), External system (Google Navigation)

Let's assume my list of usecases/requirements and my derivates above were complete I now make some conclusions (based on cgeos existing UI, because trying to make this exercise with a green-field approach is just too much for a Sunday afternoon):
- Cache Popup: From here I only need to say "go to this cache" - so at first 1 button. Then I might need to say how I want to get there. Is that usually (let's say 90-95%) the same mode? Like if I always just walk then I don't need a bike, car, .. selection. However if I 70% walk to the next cache, 30% drive to it and maybe walk the last few meters then I might need it. Let's assume the later (70/30) is the case, we then have the option of 2 buttons or 1 button + a selection or ... just brainstorming here: Can we somehow figure out how the user wants to get there with the 90-95% threshold? Like if the target is in 100m distance and in the woods 100% he'll walk there, if it's 200km away he'll use the car. If we could reach such a level of "intelligence" we could simplify the UI to have only 1 button -> value realized!
So looking at the 3 existing buttons I'd say that even today we can remove one of them - Compass/target are very similar.
- Map context menu: With the text above we can say "set as target" and "navigate" are required. For the "append/prepend" we have to look at 2 scenarios:
- No route active: "Prepend" is not visible - good! "Append to route" does something very similar to "set as target". Is there any usecase above which really requires both?
- Route active: "Append" puts it at the end of the route, "Prepend" sets it as the next target inside the route, "Set as target" sets it as next target outside the route.
In both cases "Set as target" is very similar to one of the other actions, so we could come to the conclusion to reduce it to "Go there now" and "Go there after all the other caches".

----------------------

After writing this quite long text I realize even more than before: Text is not a suitable format for this! So I'm ending this comment here, even though it's certainly not perfect, many nuances are missing, but I hope my general perspective on this comes accross.
Maybe a last quick take-away: The key question always must be: Do we really want to burden the user with the decision how to achieve that (which button to use)? This burden should only be put on him if it brings considerable value - if it's a scenario for 1% of user and it improves performance for them by only 10% it's not really worth it, because the net benefit is 0,1%!

**Reply by ztNFny** [reply]
Just thought of another example that I find very fitting to describe c:geos current problems. The above discussed problem was "there are too many buttons" - but that's not all the problem there is. Sometimes using c:geo is just very very complicated!!

Requirement: I'm looking at a map of caches and want to find the optimal route to visit them all.

Current solution (the best I know):
1. Toolbar -> More -> Show as list
2. Toolbar -> More -> Manage Caches -> Append all to route
3. Back to map
4. Open route options bottom sheet
5. Edit individual route
6. Toolbar -> Optimize route
7. Confirm I want to optimize
8. Confirm I want to use the optimization result
9. Go back to map

That's a total of 15 taps!

**Reply by fm-sys** [reply]
Just remove enough to get UI clean was a pretty stupid idea from my side, so I'm glad you turned that topic into a really valuable discussion. Well written and I agree with pretty much all you've said.

Personally I have to say that I don't like apps doing too much magic in the background, but I definitely agree that we should focus more on usecases and workflows instead of just adding features without thinking about how they are used.

**Comment by ztNFny** [comment]
> @ztNFny is privately using a "cleaned up" version of c:geo.

The stuff I removed is certainly not something that could go into c:geo, I'm listing it here because it was brought up:
- Cache menu:
  - Text to speech
  - Caches around
  - Log in browser (seriously, why does this exist!?!)
- Trackables, in CacheDetails, Log Dialog - I just don't care about that sub-game
- Live map - I do my caching offline
- log_status_marker (the colored bar on the left in cache lists)
- all texts in toolbars - icon is good enough. Also expanded the number of icons in most toolbar
- replaced the offline log menu with "log as Found"
- UnifiedMap quick launch item
- replaced tabs with icons in CacheDetailsActviity
- Map menu:
  - Elevation
  - Create UDF

**Comment by eddiemuc** [comment]
> Log in browser (seriously, why does this exist!?!)

Just because you asked: this is immensely useful if you want to log a premium cache for a secondary basic user going caching with you (e.g. your child with his/her own basic account). Because stupid gc.com does not allow that user to access the details page but just the logging page directly.
(Use Case: I am logged in c:geo as my own premium user, at the same time I am logged in the browser with my childs basic account)

**Reply by ztNFny** [reply]
So it's only relevant if 3 conditions are true: Logged-in user is Premium (edit: not true - it might also be a Basic user who got the GPX) + Cache is PMO + user is interested in log-link for basic users. The first are easy to check (and right now they are not) -> already the menu item will be gone in ~80% of cases -> https://github.com/cgeo/cgeo/pull/15442
The third would require the users input, e.g. with an option in account setup. Maybe there are other options that are also relevant in this "I'm Premium and with a basic user" that could be grouped with this to warrant such an option.

**Comment by moving-bits** [comment]
While I understand the "technical approach" discussed [further up](https://github.com/cgeo/cgeo/discussions/15434#discussioncomment-8734617), this won't help most users, IMHO - they would need to check long lists with menu entries to activate/deactivate, while they don't understand the meaning for many of them, probably don't even know where the specific menu is located. To my understanding, this approach is more suited for the very experienced user that knows most of the options by heart.

Let me scribble down a different approach (that somehow builds upon this idea) - that's what I have pitched briefly in our last dev call:

Instead of enabling/disabling menu items per one-by-one, do enable/disable larger functional blocks (and for this not only menu items, but other related visual elements as well). I can imagine not more than a handful of switches, each of which enables a certain portion of c:geo, some functionality across the whole of the app. 

"I never use routing" was an example mentioned in one of the discussions. So let's imagine, this would be one of those switches. If disabled, a couple of things would be hidden from the user, eg.:
- track/routing quick settings: load individual route and all sub-functions (edit, reload, show/hide, focus, delete, optimize); export individual route; set route start as target
- map quick settings: set route start as target
- long tap on route
- elevation chart
- long tap on map: prepend/append to route
- long tap on cache/waypoint: prepend/append to route
- settings: remove from route on log; line appearance for routing line
(probably more)

If this includes "internal routing" as well, more items could be hidden:
- map: check for missing routing data
- several settings wrt internal routing (folder, auto-download, update, profile selection
- update check for routing data
- ...

If this includes all routing options, even more would be hidden: (though I would not go thus far):
- several more settings (except those used for external navigation)
- map quick settings: profile selection
- ...

This is just an example, and probably incomplete. The story behind this example would be: "How would c:geo look, if it had no routing functionality" (or whatever functional block you want to exclude).

"Disclaimer": Don't know, though, if this approach would actually work in general, if it removes enough visual clutter, and if we were able to identify a handful of large-enough blocks (and could agree on them). Our use-cases (and those of our users) may be too diverse to find a good solution.

**Comment by Lineflyer** [comment]
I have no time currently to in deep follow the discussion but I still want to provide my personal view:

IMHO we might have some different categories of problems with the feature set in c:geo:
1. Most features are brilliant and helpful for power users and also well integrated into our UI 
2. Some features are also technically brilliant and helpful but the UI is not intuitive to use (e.g. some advanced features of the individual routing stuff) 
3. Some features have been integrated because it was technically possible but we have no idea, whether those are helpful for users (e.g. showing height information)

In general removing existing and released features (without a similar replacement) is problematic as users might rely on them.
I am no UI expert, but for me its sounds like the following workflow could help to categorize features in basic/advanced and develop ideas, whether it is possible that users can enable advanced features and only present the basic feature set upon installation. 

This feature split should be as easy as possible (meaning: Better a simple switch than another setting menu where users can decide what feature to activate).
